Split and simplify //daml-foundations/daml-ghc:daml-ghc-lib

aherrmann-da opened the pull request:
- Split the `//daml-foundations/daml-ghc:daml-ghc-lib` target into smaller targets
- Remove unused or unnecessary dependencies

This is to improve incremental compilation and give Bazel more opportunity to parallelize the build. For a simple test, changing `daml-foundations/daml-ghc/src/DA/Service/Daml/Compiler/Impl/Handle.hs` and recompiling `damlc` used to take ~150s before, and now takes ~67s on my machine.

In future we may want to further separate the targets sources into separate source directories.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

aherrmann-da commented:
To a large extent Bazel operates on the granularity of targets. Smaller targets give Bazel more opportunity to do its job, so it's good to strive for small targets.

> At some point we'll have so many small packages that becomes the problem and start trading off in the other direction.

As everything it's a trade-off and we should find a good balance. I don't think this PR is going into these extremes. But I'm happy to discuss the details, if you object to some particular change here. In particular, I'm not saying each target should only be one Haskell module. I doubt that this approach is a good fit for Haskell.

> It seems like Bazel isn't delivering the goods, so we're splitting into smaller and smaller packages. [...] Any way to make Bazel "just work"?

Surely, Bazel and rules_haskell could still be improved. E.g. having a persistent builder could give us sub-target incremental builds. However, even then there is still value in smaller targets. E.g. they also better lend themselves to remote caching. Also, I find smaller targets easier to handle.

To keep the discussion closer to the actual changes in this PR:

Looking at the module graph it seems that there were multiple targets hiding within `daml-ghc-lib` already. I think making this explicit in the targets also helps developers better understand the code base. It should also make future refactorings easier, as it's easier to change or move a smaller target.

Also, before `daml-ghc-lib` contained some test modules. I think test modules should always factored out into separate test support libraries. The final binary should not depend on such code.

```
daml-foundations/daml-ghc/src
├── DA
│   ├── Daml
│   │   └── GHC
│   │       ├── Compiler
│   │       │   ...
│   │       └── Damldoc
│   │           ...
│   │           ├── Render
│   │           │   ├── Tests.hs
│   │           │   ...
│   │           ├── Tests.hs
│   │           ...
│   ├── Service
│   │   └── Daml
│   │       ├── Compiler
│   │       │   ...
│   │       ├── LanguageServer
│   │       │   ...
│   │       └── LanguageServer.hs
│   └── Test
│       ...
└── Development
    ...
```

cocreature commented:
I tend to agree with @aherrmann-da here. The factoring looks reasonable and if it isn’t our module hierarchy isn’t reasonable either given that it’s based on that.

hurryabit commented on daml-foundations/daml-ghc/BUILD.bazel:
```diff
@@ -14,74 +14,269 @@ load(
+# This target takes a while to compile, probably due to TemplateHaskell.
+# It's separated into its own target not to inflict that compilation
+# time on changes outside the DA.Daml.GHC.Damldoc.Types module
+da_haskell_library(
```
Can we put all of damldoc into this library?

aherrmann-da replied:
The rest of `damldoc` is in [the target called `daml-ghc-damldoc`](https://github.com/digital-asset/daml/pull/1165/files#diff-48f80e007f92d8454827e7e5ea770aa8R214). I'm guessing you'd want all of `damldoc` in that target.

Sure, we can do that. However, as the comment says
```
# This target takes a while to compile, probably due to TemplateHaskell.
# It's separated into its own target not to inflict that compilation
# time on changes outside the DA.Daml.GHC.Damldoc.Types module
```

But, if you prefer to not separate out `Damldoc.Types` knowing that cost, I'm happy to keep it in there.

hurryabit replied:
Putting a single module into a package seems quite extreme to me and is something I'd like to avoid. Let's move it to the rest of of `damldoc`. We can probably improve compilation time by removing the Template Haskell.

neil-da commented:
I'm not objecting, particularly pulling tests out. I'd much rather we had separate directories for each da-haskell-library rather than using globs to select out pieces of the code. That way the humans benefit from the separation too.

That said, my experience in the past is that a single unified hierarchy as actually the best thing to use. Although doing that with Bazel seems totally infeasible.

aherrmann-da commented:
> I'd much rather we had separate directories for each da-haskell-library rather than using globs to select out pieces of the code. That way the humans benefit from the separation too.

That's fair. I can do that right away.

The Windows CI failure might be due to too long file names, so I'll probably need to pick shorter target names as well.

aherrmann-da commented:
@neil-da I've moved the new targets into separate directories with simplified target names based on the module path. PTAL
I've also updated the README. 
Seems this also fixed the Windows issue.
